### PR TITLE
Return ENOENT when reading /proc/{pid}/task of an exited process

### PR DESCRIPTION
**Pull request by zhangningdlut**
There will be a deadloop when we use getdents to read /proc/{pid}/task
of an exited process

Like this:

Process A is running
                         Process B: open /proc/{pid of A}/task
Process A exits
                         Process B: getdents /proc/{pid of A}/task

Then, process B will fall into deadloop, and return "." and ".."
in loops and never ends.

This patch returns ENOENT when use getdents to read /proc/{pid}/task
if the process is just exited.

Signed-off-by: chris.zn <chris.zn@antfin.com>

### CODE REVIEW

**Comment by zhangningdlut** [comment]
I signed it

> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogle%2Fgvisor%2Fpull%2F282) for more info**.

**Comment by nlacasse** [comment]
Thanks for this PR!

The CLA checker is complaining that you have not signed the CLA.  Did you maybe use a different email address to sign?  It must match your GitHub email address.

**Comment by zhangningdlut** [comment]
> Thanks for this PR!
> 
> The CLA checker is complaining that you have not signed the CLA. Did you maybe use a different email address to sign? It must match your GitHub email address.

I from antfin, I thought I have already signed.
You means I should sign it by personal ?

**Comment by amscanne** [comment]
The bot doesn't know it for some reason. I believe your email can be covered under corporate CLA, but you need to use a googlegroup to include the addresses that should be covered. (I'm not sure, but I'm interpreting the instructions @ https://cla.developers.google.com/clas).

**Comment by avagin** [comment]
Could you add a test for this? I think you can add a new test case in https://github.com/google/gvisor/blob/master/test/syscalls/linux/proc.cc.

you can use this bazel command to run the test:
blaze test test/syscalls:proc_test_native test/syscalls:proc_test_runsc_ptrace

**Comment by zhangningdlut** [comment]
> Could you add a test for this? I think you can add a new test case in https://github.com/google/gvisor/blob/master/test/syscalls/linux/proc.cc.
> 
> you can use this bazel command to run the test:
> blaze test test/syscalls:proc_test_native test/syscalls:proc_test_runsc_ptrace

Hi, I have pushed the test :)

**Comment by zhangningdlut** [comment]
> The bot doesn't know it for some reason. I believe your email can be covered under corporate CLA, but you need to use a googlegroup to include the addresses that should be covered. (I'm not sure, but I'm interpreting the instructions @ https://cla.developers.google.com/clas).

Thanks !
Have I signed the CLA?
I add my email to the Agreement.

File: test/syscalls/linux/proc.cc [hunk 1]
```diff
@@ -1908,6 +1908,23 @@ void CheckDuplicatesRecursively(std::string path) {
 
 TEST(Proc, NoDuplicates) { CheckDuplicatesRecursively("/proc"); }
 
+TEST(Proc, ReadDir) {
+  int child_pid = fork();
+  ASSERT_GE(child_pid, 0);
```
Comment (avagin) [inline test/syscalls/linux/proc.cc]: ASSERT_THAT(child_pid, SyscallSucceeds());

File: test/syscalls/linux/proc.cc [hunk 2]
```diff
@@ -1908,6 +1908,23 @@ void CheckDuplicatesRecursively(std::string path) {
...
+  if (child_pid == 0) {
+    while(1){};
+  }
+  else
+  {
+    char name[100];
+    sprintf(name, "/proc/%d/task", child_pid);
+    DIR *dir = opendir(name);
+    ASSERT_THAT(kill(child_pid, SIGKILL), SyscallSucceeds());
+    wait(NULL);
```
Comment (avagin) [inline test/syscalls/linux/proc.cc]: EXPECT_THAT(waitpid(pid, NULL, 0), SyscallSucceedsWithValue(pid));

File: test/syscalls/linux/proc.cc [hunk 3]
```diff
@@ -1908,6 +1908,23 @@ void CheckDuplicatesRecursively(std::string path) {
...
+    ASSERT_EQ((char *)readdir(dir), (char*)NULL);
```
Comment (avagin) [inline test/syscalls/linux/proc.cc]: You need to check that a proper errno is returned

File: test/syscalls/linux/proc.cc [hunk 4]
```diff
@@ -1908,6 +1908,23 @@ void CheckDuplicatesRecursively(std::string path) {
 
 TEST(Proc, NoDuplicates) { CheckDuplicatesRecursively("/proc"); }
 
+TEST(Proc, ReadDir) {
+  int child_pid = fork();
+  ASSERT_GE(child_pid, 0);
+  if (child_pid == 0) {
+    while(1){};
```
Comment (avagin) [inline test/syscalls/linux/proc.cc]: We don't want to heat the Earth, pls avoid busy-loops. while (1) { sleep(1000); }

File: test/syscalls/linux/proc.cc [hunk 5]
```diff
@@ -1908,6 +1908,23 @@ void CheckDuplicatesRecursively(std::string path) {
...
+  }
+  else
```
Comment (avagin) [inline test/syscalls/linux/proc.cc]: you can remove this else

**Comment by zhangningdlut** [comment]
Hi @avagin, can this patch be merged?

**Pull request merged**